### PR TITLE
For strategies.yaml, add peering_ring with go_direct, and Au tests for peering_ring.

### PR DESCRIPTION
**Pull request by ywkaras**
Caching is disabled when the next hop is a peer (in the first host group).

### CODE REVIEW

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
Making this a draft for now.  John may be adding an option to strategies.yaml that will make the TS_TEST_LOCALHOST environment variable unnecessary.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
This includes an addition to strategies.yaml, adding an optional **self** key to the **failover** map for a peering ring, whose value can specify a hostname from the peer group as identifying the local host.

**Comment by bryancall** [comment]
@jrushford or @rob05c can you please take a look at this.

**Comment by ywkaras** [comment]
[approve ci autest]

File: proxy/http/remap/NextHopSelectionStrategy.cc [hunk 1]
```diff
@@ -63,6 +63,8 @@ bool
 NextHopSelectionStrategy::Init(ts::Yaml::Map &n)
 {
   NH_Debug(NH_DEBUG_TAG, "calling Init()");
+  std::string self_host;
```
Comment (jrushford) [inline proxy/http/remap/NextHopSelectionStrategy.cc:66]: @ywkaras Consider adding this new field to the unit testing, see proxy/remap/http/unit-tests/test_NextHopConsistentHash.cc

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by jrushford** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
The autest failure is looks repeatable, always happens for tls_conn_timeout test case. But it passes in my development environment.

**Comment by ywkaras** [comment]
[approve ci RAT]

**Comment by jrushford** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci Clang-Analyzer]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
Converted this to draft until I can get Au tests on it.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

File: proxy/http/remap/NextHopSelectionStrategy.cc [hunk 2]
```diff
@@ -108,7 +110,12 @@ NextHopSelectionStrategy::Init(ts::Yaml::Map &n)
         } else if (ring_mode_val == exhaust_rings) {
           ring_mode = NH_EXHAUST_RING;
         } else if (ring_mode_val == peering_rings) {
-          ring_mode = NH_PEERING_RING;
+          ring_mode            = NH_PEERING_RING;
+          YAML::Node self_node = failover_node["self"];
+          if (self_node) {
```
Comment (jrushford) [inline proxy/http/remap/NextHopSelectionStrategy.cc:115]: @ywkaras Is this correct, `self` is part of the failover node`.  I'm not sure how this works in identifying the self host.  Maybe I'm missing something.  In testing, I'm looping when the self host is selected

File: proxy/http/remap/NextHopSelectionStrategy.cc [hunk 3]
```diff
@@ -214,8 +221,13 @@ NextHopSelectionStrategy::Init(ts::Yaml::Map &n)
               std::shared_ptr<HostRecord> host_rec = std::make_shared<HostRecord>(hosts_list[hst].as<HostRecord>());
               host_rec->group_index                = grp;
               host_rec->host_index                 = hst;
-              if (mach->is_self(host_rec->hostname)) {
+              if ((self_host == host_rec->hostname) || mach->is_self(host_rec->hostname.c_str())) {
+                if (grp != 0) {
+                  throw std::invalid_argument("self host (" + self_host + ") can only appear in first host group");
+                }
                 h_stat.setHostStatus(host_rec->hostname.c_str(), TSHostStatus::TS_HOST_STATUS_DOWN, 0, Reason::SELF_DETECT);
```
Comment (jrushford) [inline proxy/http/remap/NextHopSelectionStrategy.cc:228]: @ywkaras the self detection doesn't seem to work now.  Host status no-longer has a SELF_DETECT down for the self_host.  I'm using the peering ring mode.

**Review by jrushford — CHANGES_REQUESTED** [review]

**Review by jrushford — APPROVED** [review]
Looks good and tests ok

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci FreeBSD]

**Pull request merged**